File: packages/python-sdk/src/urisocial/resources/canvas_editor.py
"""Canvas editor resource for URI Social SDK"""

from typing import Dict, Any, List, Optional
from ..http_client import HTTPClient


class CanvasEditorResource:
    """Canvas editor operations - layered document editing"""

    def __init__(self, http: HTTPClient):
        self._http = http

    def get_draft_document(self, draft_id: str) -> Dict[str, Any]:
        """
        Get layered document for a draft

        Returns complete layered JSON document with all layers,
        ready for canvas editor rendering.

        Args:
            draft_id: Draft ID

        Returns:
            Layered document object
        """
        return self._http.get(f"/canvas-editor/drafts/{draft_id}/document")

    def update_layer(
        self,
        draft_id: str,
        layer_id: str,
        updates: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Update individual layer properties

        Args:
            draft_id: Draft ID
            layer_id: Layer ID
            updates: Dictionary of properties to update (x, y, width, height, rotation, opacity, etc.)

        Returns:
            Updated document object
        """
        return self._http.post(
            f"/canvas-editor/drafts/{draft_id}/layers/{layer_id}/update",
            json={"layer_id": layer_id, "updates": updates},
        )

    def undo(self, draft_id: str) -> Dict[str, Any]:
        """
        Undo last edit operation

        Args:
            draft_id: Draft ID

        Returns:
            Document state after undo
        """
        return self._http.post(f"/canvas-editor/drafts/{draft_id}/undo")

    def redo(self, draft_id: str) -> Dict[str, Any]:
        """
        Redo previously undone operation

        Args:
            draft_id: Draft ID

        Returns:
            Document state after redo
        """
        return self._http.post(f"/canvas-editor/drafts/{draft_id}/redo")

    def render_document(
        self,
        draft_id: str,
        aspect_ratio: str = "1:1",
        output_format: str = "png",
        quality: int = 95,
    ) -> Dict[str, Any]:
        """
        Render layered document to image

        Args:
            draft_id: Draft ID
            aspect_ratio: Target aspect ratio (default: '1:1')
            output_format: Output format ('png' or 'jpg', default: 'png')
            quality: Image quality 1-100 (default: 95)

        Returns:
            Rendered image URL
        """
        return self._http.post(
            f"/canvas-editor/drafts/{draft_id}/render",
            json={
                "aspect_ratio": aspect_ratio,
                "output_format": output_format,
                "quality": quality,
            },
        )

    def reorder_layers(self, draft_id: str, layer_order: List[str]) -> Dict[str, Any]:
        """
        Reorder layers (z-index)

        Args:
            draft_id: Draft ID
            layer_order: List of layer IDs in desired order (bottom to top)

        Returns:
            Updated document object
        """
        return self._http.post(
            f"/canvas-editor/drafts/{draft_id}/layers/reorder",
            json={"layer_order": layer_order},
        )

    def delete_layer(self, draft_id: str, layer_id: str) -> Dict[str, Any]:
        """
        Delete layer from document

        Args:
            draft_id: Draft ID
            layer_id: Layer ID to delete

        Returns:
            Updated document object
        """
        return self._http.delete(f"/canvas-editor/drafts/{draft_id}/layers/{layer_id}")

    def get_edit_history(self, draft_id: str) -> Dict[str, Any]:
        """
        Get edit history for draft

        Returns chronological list of all edit operations.

        Args:
            draft_id: Draft ID

        Returns:
            Edit history with operations, timestamps, and user info
        """
        return self._http.get(f"/canvas-editor/drafts/{draft_id}/edit-history")

File: packages/typescript-sdk/src/resources/canvas-editor.ts
import { HTTPClient } from '../client/http';

// ── Types ──────────────────────────────────────────────────────────────

export interface LayeredDocument {
  draft_id: string;
  layers: Layer[];
  canvas_width: number;
  canvas_height: number;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface Layer {
  layer_id: string;
  type: 'text' | 'image' | 'shape' | 'background';
  x: number;
  y: number;
  width: number;
  height: number;
  rotation?: number;
  opacity?: number;
  z_index: number;
  properties: Record<string, any>;
}

export interface UpdateLayerRequest {
  layer_id: string;
  updates: Partial<Layer>;
}

export interface RenderDocumentRequest {
  aspect_ratio?: string;
  output_format?: 'png' | 'jpg';
  quality?: number;
}

export interface ReorderLayersRequest {
  layer_order: string[];
}

export interface EditHistoryEntry {
  operation: string;
  timestamp: string;
  user_id: string;
  changes: Record<string, any>;
}

// ── Resource ───────────────────────────────────────────────────────────

export class CanvasEditorResource {
  constructor(private http: HTTPClient) {}

  /**
   * Get layered document for a draft
   *
   * Returns complete layered JSON document with all layers,
   * ready for canvas editor rendering.
   */
  async getDraftDocument(draftId: string): Promise<{ document: LayeredDocument }> {
    return this.http.get<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/document`);
  }

  /**
   * Update individual layer properties
   */
  async updateLayer(draftId: string, layerId: string, updates: Partial<Layer>): Promise<{ document: LayeredDocument }> {
    return this.http.post<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/layers/${layerId}/update`, {
      layer_id: layerId,
      updates,
    });
  }

  /**
   * Undo last edit operation
   */
  async undo(draftId: string): Promise<{ document: LayeredDocument }> {
    return this.http.post<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/undo`);
  }

  /**
   * Redo previously undone operation
   */
  async redo(draftId: string): Promise<{ document: LayeredDocument }> {
    return this.http.post<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/redo`);
  }

  /**
   * Render layered document to image
   */
  async renderDocument(draftId: string, request?: RenderDocumentRequest): Promise<{ image_url: string }> {
    return this.http.post<{ image_url: string }>(`/canvas-editor/drafts/${draftId}/render`, request || {});
  }

  /**
   * Reorder layers (z-index)
   */
  async reorderLayers(draftId: string, layerOrder: string[]): Promise<{ document: LayeredDocument }> {
    return this.http.post<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/layers/reorder`, {
      layer_order: layerOrder,
    });
  }

  /**
   * Delete layer from document
   */
  async deleteLayer(draftId: string, layerId: string): Promise<{ document: LayeredDocument }> {
    return this.http.delete<{ document: LayeredDocument }>(`/canvas-editor/drafts/${draftId}/layers/${layerId}`);
  }

  /**
   * Get edit history for draft
   *
   * Returns chronological list of all edit operations.
   */
  async getEditHistory(draftId: string): Promise<{ history: EditHistoryEntry[] }> {
    return this.http.get<{ history: EditHistoryEntry[] }>(`/canvas-editor/drafts/${draftId}/edit-history`);
  }
}
